### PR TITLE
Telemetry: chat-question:submitted

### PR DESCRIPTION
**Pull request by abeatrix**
Fix: https://github.com/sourcegraph/cody/issues/2916

Add logging of `chat-question:submitted` event to `handleUserMessageSubmission()` method.

Currently, we are only logging the aforementioned event in the `executeChat()` function which is only used by commands and code-action, so the event was not logged for chat questions submitted from the webview. This PR moves the logging to the beginning of the `handleUserMessageSubmission()` method, which is used by all chat-questions submitted. At the end, it will log the `chat-question:executed` event if the chat-question request was completed without error.

Other changes:
- The source is now passed to the SimpleChatPanelProvider's handleUserMessageSubmission method and logged in telemetry. This will allow better tracking of where chat requests originate from in the future.

## Test plan

Updated all the E2E tests to ensure the `CodyVSCodeExtension:chat-question:submitted ` is fired when chat question is submitted in different places, and must log before `CodyVSCodeExtension:chat-question:executed ` .

Shout out to Aditya for adding this!

## Examples

https://github.com/sourcegraph/cody/assets/68532117/489f45e9-8bf8-48ea-afad-1ed92a82b103

chat-question events from log when submitting a chat via chat view:

```
█ logEvent (telemetry disabled): CodyVSCodeExtension:chat-question:submitted VSCode {"properties":{"requestID":"bf7d4551-0ea3-44c8-b4ad-a5fd3d2f271d","chatModel":"anthropic/claude-2.0","source":"chat","traceId":"434958bf00d00a0f29b67ed4556c76f8"}}
█ telemetry-v2: recordEvent: cody.chat-question/submitted: {"parameters":{"version":0,"privateMetadata":{"requestID":"bf7d4551-0ea3-44c8-b4ad-a5fd3d2f271d","chatModel":"anthropic/claude-2.0","source":"chat","traceId":"434958bf00d00a0f29b67ed4556c76f8"}},"timestamp":"2024-02-15T16:35:23.640Z"}
█ logEvent (telemetry disabled): CodyVSCodeExtension:chat-question:executed VSCode {"properties":{"requestID":"bf7d4551-0ea3-44c8-b4ad-a5fd3d2f271d","chatModel":"anthropic/claude-2.0","source":"chat","traceId":"434958bf00d00a0f29b67ed4556c76f8","contextSummary":{"search":15}},"opts":{"hasV2Event":true}}
█ telemetry-v2: recordEvent: cody.chat-question/executed: {"parameters":{"version":0,"metadata":[{"key":"search","value":15},{"key":"recordsPrivateMetadataTranscript","value":1}],"privateMetadata":{"properties":{"requestID":"bf7d4551-0ea3-44c8-b4ad-a5fd3d2f271d","chatModel":"anthropic/claude-2.0","source":"chat","traceId":"434958bf00d00a0f29b67ed4556c76f8","contextSummary":{"search":15}},"promptText":"testing"}},"timestamp":"2024-02-15T16:35:26.705Z"}
```

chat-question events from log when submitting a chat via menu:

```
█ logEvent (telemetry disabled): CodyVSCodeExtension:chat-question:submitted VSCode {"properties":{"requestID":"8f70b332-79a2-431f-bd7c-397d89962c73","chatModel":"anthropic/claude-2.0","source":"menu","traceId":"5d3bacc004ca95bfb74ac5dde836e0b7"}}
█ telemetry-v2: recordEvent: cody.chat-question/submitted: {"parameters":{"version":0,"privateMetadata":{"requestID":"8f70b332-79a2-431f-bd7c-397d89962c73","chatModel":"anthropic/claude-2.0","source":"menu","traceId":"5d3bacc004ca95bfb74ac5dde836e0b7"}},"timestamp":"2024-02-15T16:35:46.132Z"}
█ logEvent (telemetry disabled): CodyVSCodeExtension:chat-question:executed VSCode {"properties":{"requestID":"8f70b332-79a2-431f-bd7c-397d89962c73","chatModel":"anthropic/claude-2.0","source":"menu","traceId":"5d3bacc004ca95bfb74ac5dde836e0b7","contextSummary":{"search":8}},"opts":{"hasV2Event":true}}
█ telemetry-v2: recordEvent: cody.chat-question/executed: {"parameters":{"version":0,"metadata":[{"key":"search","value":8},{"key":"recordsPrivateMetadataTranscript","value":1}],"privateMetadata":{"properties":{"requestID":"8f70b332-79a2-431f-bd7c-397d89962c73","chatModel":"anthropic/claude-2.0","source":"menu","traceId":"5d3bacc004ca95bfb74ac5dde836e0b7","contextSummary":{"search":8}},"promptText":"hello"}},"timestamp":"2024-02-15T16:35:47.993Z"}
```

### CODE REVIEW

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts [hunk 1]
```diff
@@ -383,9 +385,21 @@ export class SimpleChatPanelProvider implements vscode.Disposable, ChatSession {
...
+            const sharedProperties = {
+                requestID,
+                chatModel: this.chatModel.modelID,
+                source,
+                traceId: span.spanContext().traceId,
+            }
+            telemetryService.log('CodyVSCodeExtension:chat-question:submitted', sharedProperties)
+            telemetryRecorder.recordEvent('cody.chat-question', 'submitted', {
+                privateMetadata: sharedProperties,
+            })
```
Comment (akalia25) [inline vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:401]: Hey @abeatrix great PR, thanks for taking this on!!

For `privateMetadata` we only log this for dotcom users, so we will lose all this data for customers. We can add the fields that are not free-form text to `metadata` and have collection of that data everyone.

Also I think we can include logging of transcript within v2 telemetry here.

Something like this?: We can add it to `metadata` as long as its either a numeric value or a enum :) -- I didn't check if source, and modelID are all enums but if they are we can do this :) 

```
telemetryRecorder.recordEvent('cody.chat-question', 'submitted', {
                metadata: {
                      chatModel: this.chatModel.modelID,
                      source:source
                      recordsPrivateMetadataTranscript:
                                      authStatus.endpoint && isDotCom(authStatus.endpoint) ? 1 : 0,
                privateMetadata: { 
                      promptText: inputText
                }
                
            })
```

File: vscode/test/e2e/chat-input.test.ts [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
+        'CodyInstalled',
+        'CodyVSCodeExtension:auth:clickOtherSignInOptions',
+        'CodyVSCodeExtension:login:clicked',
+        'CodyVSCodeExtension:auth:selectSigninMenu',
+        'CodyVSCodeExtension:auth:fromToken',
+        'CodyVSCodeExtension:Auth:connected',
+        'CodyVSCodeExtension:chat-question:submitted',
+        'CodyVSCodeExtension:chat-question:executed',
+    ],
+})('editing messages in the chat input', async ({ page, sidebar }) => {
```
Comment (akalia25) [inline vscode/test/e2e/chat-input.test.ts:18]: 🔥 🔥 thank you for adding this and making our logging more stable!!

**Review by chenkc805 — APPROVED** [review]

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts (hunk 1 above)
Comment (abeatrix) [reply vscode/src/chat/chat-view/SimpleChatPanelProvider.ts]: Thank you! I just updated the PR to include your suggestion. Let me know if that looks off to you :)

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts [hunk 3]
```diff
@@ -383,9 +385,35 @@ export class SimpleChatPanelProvider implements vscode.Disposable, ChatSession {
...
+            telemetryService.log('CodyVSCodeExtension:chat-question:submitted', sharedProperties)
+            telemetryRecorder.recordEvent('cody.chat-question', 'submitted', {
+                metadata: {
+                    // Flag indicating this is a transcript event to go through ML data pipeline. Only for DotCom users
+                    // See https://github.com/sourcegraph/sourcegraph/pull/59524
+                    recordsPrivateMetadataTranscript:
+                        authStatus.endpoint && isDotCom(authStatus.endpoint) ? 1 : 0,
+                },
+                privateMetadata: {
+                    ...sharedProperties,
```
Comment (akalia25) [inline vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:408]: One last thing, I think `...sharedProperties` should be on `metadata` :).

Tested with `...sharedProperties` on metadata and sample payload below :) 

Tested, and 
```
recordEvent: cody.chat-question/submitted: {"parameters":{"version":0,"metadata":[{"key":"recordsPrivateMetadataTranscript","value":0},{"key":"requestID","value":"23a2958e-0820-40bb-a218-cf5ded7e8d8d"},{"key":"chatModel","value":"anthropic/claude-2"},{"key":"source","value":"chat"},{"key":"traceId","value":"00000000000000000000000000000000"}],"privateMetadata":{}},"timestamp":"2024-02-15T17:54:10.529Z"}```

Comment (akalia25) [reply vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:408]: Actually never mind, `metadata` is still giving issues with string types that are enum. I'll follow-up on this separately but we're good to merge this as is :)

**Review by akalia25 — APPROVED** [review]
LGTM! Thank you for adding this!! 🚀

**Pull request merged**